cdc: store information on ttl in "ttl" column, not in tuples

piodul opened the pull request:
This patch changes the way TTL is stored in the CDC log table. Instead of including TTL of cell `X` in the third element of the tuple in column `_X`, TTL is written to the previously unused column `ttl`. This is done for cosmetic purposes.

This implementation works under assumption that there will be only one TTL included in a mutation coming from a CQL write. This might not be the case when writing a batch that modifies the same row twice, e.g.:

```
BATCH
INSERT INTO ks.t (pk, ck, v1) VALUES (1,2,3) USING TTL 10;
INSERT INTO ks.t (pk, ck, v2) VALUES (1,2,3) USING TTL 20;
END BATCH
```

In this case, this implementation will choose only one TTL value to be written in the CDC log:

```
... | batch_seq_no | _ck | _pk | _v1    | _v2    | operation | ttl
...-+--------------+-----+-----+--------+--------+-----------+-----
... |            0 |   2 |   1 | (0, 3) | (0, 3) |         1 |  20
```

This behavior might be changed as a part of issue #5719, which considers splitting a batch write mutation when it contains multiple writes to the same row.

Fixes #5689
Tests: unit(dev)

nyh commented:
The text of the commit message is self-contradictory - you say:

> This implementation **assumes** that, in a single write, if two cells from the same row have TTL set, then those TTLs should be equal. **Otherwise**, only one of those TTLs will be chosen and written into the CDC log.

If it assumes, how can there be an otherwise?

Looking at the discussion in https://github.com/scylladb/scylla/issues/5689, it seems it was suggested (though I'm not 100% confident... is there really no way?) that with real CQL writes - not programatic mutations from repairs, etc. - we'll really have just one TTL per mutation. Even with a batch, there will be multiple mutations each with just what TTL. Is this what we are assuming? If so, probably we should be explicit about this. There would be no "otherwise" in real CQL writes. Perhaps you can just say that in the unsupported case of fake/programatic/repair writes there can be such differing TTLs  and we use one of them.

All that being said, I don't know why we want to do this change at all. What was wrong with having a TTL per cell? Are we trying to save space (when there are TLLs)?

haaawk commented:
> The text of the commit message is self-contradictory - you say:
> 
> > This implementation **assumes** that, in a single write, if two cells from the same row have TTL set, then those TTLs should be equal. **Otherwise**, only one of those TTLs will be chosen and written into the CDC log.
> 
> If it assumes, how can there be an otherwise?
> 
> Looking at the discussion in #5689, it seems it was suggested (though I'm not 100% confident... is there really no way?) that with real CQL writes - not programatic mutations from repairs, etc. - we'll really have just one TTL per mutation. Even with a batch, there will be multiple mutations each with just what TTL. Is this what we are assuming? If so, probably we should be explicit about this. There would be no "otherwise" in real CQL writes. Perhaps you can just say that in the unsupported case of fake/programatic/repair writes there can be such differing TTLs and we use one of them.
> 
> All that being said, I don't know why we want to do this change at all. What was wrong with having a TTL per cell? Are we trying to save space (when there are TLLs)?

@asias Could you comment on repair/streaming having mutations with cells that have different timestamps/ttls please?

kbr- commented:
@haaawk 
repair/streaming doesn't interact with CDC. CDC mutations are added during write coordination.

haaawk commented:
> The text of the commit message is self-contradictory - you say:
> 
> > This implementation **assumes** that, in a single write, if two cells from the same row have TTL set, then those TTLs should be equal. **Otherwise**, only one of those TTLs will be chosen and written into the CDC log.
> 
> If it assumes, how can there be an otherwise?
> 
> Looking at the discussion in #5689, it seems it was suggested (though I'm not 100% confident... is there really no way?) that with real CQL writes - not programatic mutations from repairs, etc. - we'll really have just one TTL per mutation. Even with a batch, there will be multiple mutations each with just what TTL. Is this what we are assuming? If so, probably we should be explicit about this. There would be no "otherwise" in real CQL writes. Perhaps you can just say that in the unsupported case of fake/programatic/repair writes there can be such differing TTLs and we use one of them.
> 
> All that being said, I don't know why we want to do this change at all. What was wrong with having a TTL per cell? Are we trying to save space (when there are TLLs)?

@asias Could you comment on repair/streaming having mutations with cells that have different timestamps/ttls please?

> > The text of the commit message is self-contradictory - you say:
> > > This implementation **assumes** that, in a single write, if two cells from the same row have TTL set, then those TTLs should be equal. **Otherwise**, only one of those TTLs will be chosen and written into the CDC log.
> > 
> > 
> > If it assumes, how can there be an otherwise?
> > Looking at the discussion in #5689, it seems it was suggested (though I'm not 100% confident... is there really no way?) that with real CQL writes - not programatic mutations from repairs, etc. - we'll really have just one TTL per mutation. Even with a batch, there will be multiple mutations each with just what TTL. Is this what we are assuming? If so, probably we should be explicit about this. There would be no "otherwise" in real CQL writes. Perhaps you can just say that in the unsupported case of fake/programatic/repair writes there can be such differing TTLs and we use one of them.
> > All that being said, I don't know why we want to do this change at all. What was wrong with having a TTL per cell? Are we trying to save space (when there are TLLs)?
> 
> @asias Could you comment on repair/streaming having mutations with cells that have different timestamps/ttls please?

On a s

haaawk commented:
> @haaawk
> repair/streaming doesn't interact with CDC. CDC mutations are added during write coordination.

Right.

haaawk commented:
@nyh We decided during this week CDC call to drop ttl per cell and keep only ttl per row. At the moment ttl for row is not used.

nyh commented:
@haaawk yes, I guess it is fine, but it took me a while to figure out why and I think it should be spelled out more clearly in the commit message why it is fine - i.e., that real CQL writes will always have just a single TTL, not mixed TTLs. Once I realized this, I understand why you did this change, but I had to realize this myself, as it wasn't explicitly explained in the commit message.

haaawk commented:
I agree with your comment about a better explanation @nyh

piodul commented:
@haaawk  @kbr- @nyh 

Updated the commit description - I hope it is more clear now.

Due to a bug, ttl was not written when `preimage` option was switched on - I fixed that and extended tests to check for that scenario.

kbr- commented on cdc/log.cc:
```diff
@@ -548,22 +555,29 @@ class transformer final {
                 };
 
+                const auto& marker = r.row().marker();
+                if (marker.is_expiring()) {
```
The comment on `row_marker::is_expiring()` says:
```
    // Can be called only when is_live().
```
but there are cases here when `!marker.is_live()`, aren't there?

piodul replied:
Added a check for liveness.

psarna commented on cdc/log.cc:
```diff
@@ -548,22 +555,29 @@ class transformer final {
                                 assert(pikey->explode() != log_ck.explode());
                                 res.set_cell(*pikey, *dst, atomic_cell::make_live(*dst->type, ts, tuple_type_impl::build_value(values)));
                             }
+
```
I think that one got here by accident

piodul replied:
Right, fixed.

psarna commented on test/boost/cdc_test.cc:
```diff
@@ -307,6 +316,8 @@ SEASTAR_THREAD_TEST_CASE(test_primary_key_logging) {
         // UPDATE ks.tbl SET val = 555 WHERE pk = 2 AND pk2 = 11 AND ck = 111 AND ck2 = 1111
         assert_row(2, 11, 111, 1111);
+        // UPDATE ks.tbl USING TTL 3600 SET val = 444 WHERE pk = 3 AND pk2 = 11 AND ck = 111 AND ck2 = 1111
+        assert_row(3, 11, 111, 1111, 3600);
```
hm, do I misremember how row markers and ttls work? I was under the impression that this kind of `UPDATE` will only update the TTL for changed columns, without touching the row marker (which will only be set if data is inserted with `INSERT`). If that's the case, shouldn't `ttl` field in the cdc table remain untouched here? I saw in the `transformer` code that this column is only responsible for keeping the row ttl, i.e. the one kept in the row marker, leaving per-cell ttls to be kept in `res` (whatever it's used for).
It's very likely that there's no bug here at all and I just misunderstand how this code is supposed to work, but in any case, please clarify for the sake of my education :)

piodul replied:
We would like to use the `ttl` column for TTLs that are set both by INSERTs and UPDATEs. If you follow the logic closely in `transformer::transform` you will see that `row_ttl` will contain either ttl of the row marker, or from a cell. I see now that `row_ttl` is a bad name and why it is confusing, so I changed it to a more neutral `ttl`.

psarna replied:
👌

psarna commented:
Thanks, queued to /next. I'm not using this treacherous author-changing `squash and merge` button ever again, so this PR may need to be closed manually after /next is promoted.

piodul commented:
I can see it landed on master, so I'll close this pull request.
